[docs] new Ray Cluster documentation (#13839)
ray-project/ray: doc/source/cluster/index.rst

Distributed Ray Overview
========================

> What is Distributed Ray?

Reviewer: Hmm, why do we want to coin a new term called "Distributed Ray"? Why not just call it "Ray Cluster", so that it is not a new concept?

Otherwise, we should not capitalize "Distributed" everywhere in this documentation

Author: I changed it to "What is a Ray cluster?"

    To connect to this Ray runtime from another node, run
      ray start --address='<ip address>:6379' --redis-password='<password>'

  If connection fails, check your firewall settings and network configuration.

The command will print out the address of the Redis server that was started
(the local node IP address plus the port number you specified).

**Then on each of the other nodes**, run the following. Make sure to replace
``<address>`` with the value printed by the command on the head node (it
should look something like ``123.45.67.89:6379``).

Note that if your compute nodes are on their own subnetwork with Network
Address Translation, to connect from a regular machine outside that subnetwork,
the command printed by the head node will not work. You need to find the
address that will reach the head node from the second machine. If the head node
has a domain address like compute04.berkeley.edu, you can simply use that in
place of an IP address and rely on the DNS.

.. code-block:: bash

  $ ray start --address=<address> --redis-password='<password>'
  --------------------
  Ray runtime started.
  --------------------

  To terminate the Ray runtime, run
    ray stop

> 

Reviewer: Are there any instructions anywhere for starting Ray clusters manually? Where have the manual Ray start commands been moved to?

This is important to keep somewhere because there will still be a subset of users that cannot use the Ray cluster launcher but need to start a ray cluster.

> One of Ray's strengths is the ability to leverage multiple machines in the same program. Ray can, of course, be run on a single machine (and is done so often) but the real power is using Ray on a cluster of machines.

Reviewer: ```suggestion
One of Ray's strengths is the ability to leverage multiple machines in the same program. Ray can, of course, be run on a single machine (and is done so often), but the real power is using Ray on a cluster of machines.
```

> Distributed Ray is powered by the Ray Cluster Launcher, which automatically provisions machines and launches a multi-node Ray cluster. You can use the cluster launcher on GCP, Amazon EC2, Azure, or even Kubernetes. Distributed Ray also makes use of the Ray Autoscaler, which gives Ray the ability "auto-scale," meaning that it can interact with a Cloud Provider to request or release instances according to application workload.

Reviewer: This paragraph is a bit misleading; see changes below:

```suggestion
You can use the Ray Cluster Launcher to provision machines and launch a multi-node Ray cluster. You can use the cluster launcher on GCP, Amazon EC2, Azure, or even Kubernetes. Ray clusters can also make use of the Ray Autoscaler, which allows Ray to interact with a cloud provider to request or release instances according to application workload.
```

> How does it work?

Reviewer: "How does it work" should not come before "Supported Features" - can you please move this paragraph down to the bottom of the page? 

The below paragraphs are details that reference a bunch of new concepts.

Author: The paragraph about supported features was a bit confusing, so I removed it altogether

> The Ray Cluster Launcher will automatically enable a load-based autoscaler. The autoscaler resource demand scheduler will look at the task, actor, and placement group resource demands from the cluster, and try to add the minimum list of nodes that can fulfill these demands. When nodes are idle for more than :ref:`idle_timeout_minutes <cluster-configuration-idle-timeout-minutes>`, they will be removed. The head node is never removed.

Reviewer 2: ```suggestion
The Ray Cluster Launcher will automatically enable a load-based autoscaler. The autoscaler resource demand scheduler will look at the pending tasks, actors, and placement groups from the cluster, and try to add nodes that can fulfill these demands. When nodes are idle for more than :ref:`idle_timeout_minutes <cluster-configuration-idle-timeout-minutes>`, they will be removed. The head node is never removed.
```

IMO "resource demands" is actually kinda confusing. Also, we don't want to guarantee these policy related minimums since they're not necessarily true or a good idea.

Author: Added to the commit

> * **Placement groups**: :ref:`Placement groups <ray-placement-group-doc-ref>` allow users to atomically reserve groups of resources across multiple nodes (i.e., gang scheduling).

Reviewer 2: I know we call this a feature during development, but is this really a feature? I feel like it's table stakes, and it's implied that it exists.

Reviewer: This isn't really specific to Ray clusters too right? Maybe this belongs more so in a usage guide.

Author: I'll delete the section since it's a bit random
